### PR TITLE
avoid generalization inside of aliases

### LINKED ISSUE #4: Ambiguity for projection candidates that aren't equal, but *are* equal modulo normalization
*edit: after https://github.com/rust-lang/rust/pull/119106 this only affects higher ranked aliases and should not be an issue in practice*

```rust
fn needs_iter(_: impl Iterator<Item = i32>) {}

fn main() {
    needs_iter(IntoIterator::into_iter([0i32; 32]))
}
```

Bails out with two ambiguous candidates:

```
Candidate { source: Impl(DefId(2:3663 ~ core[523c]::array::iter::{impl#2})), result: Canonical { value: Response { certainty: Yes, var_values: CanonicalVarValues { var_values: [i32] }, external_constraints: ExternalConstraints(Interned(ExternalConstraintsData { region_constraints: QueryRegionConstraints { outlives: [], member_constraints: [] }, opaque_types: [] }, PrivateZst)) }, max_universe: U0, variables: [] } },
  Candidate { source: AliasBound, result: Canonical { value: Response { certainty: Yes, var_values: CanonicalVarValues { var_values: [<[i32; 32] as std::iter::IntoIterator>::Item] }, external_constraints: ExternalConstraints(Interned(ExternalConstraintsData { region_constraints: QueryRegionConstraints { outlives: [], member_constraints: [] }, opaque_types: [] }, PrivateZst)) }, max_universe: U0, variables: [] } }]
```

That is, we have that `<[i32; 32] as IntoIterator>::Item` normalizes to either `i32` or `<[i32; 32] as std::iter::IntoIterator>::Item` via an impl or alias bound, respectively.

### LINKED ISSUE #79: astconv/typeck unnecessarily instantiates type variables in non-root universe
- #19 
- #60

---

When instantiating substs for ADTs, confirming methods, etc., we create new type variables in the current universe, rather than the root universe, even though these variables could be in the root universe since they represent the "surface" types of a Rust program.

This causes us to unnecessarily generalize aliases, leading to ambiguity.

---

TODO: minimize this test case:
```rust
use syn::Ident;
use syn::Attribute;
use syn::parse::*;

struct TypeParam {
    pub attrs: Vec<Attribute>,
    pub ident: Ident,
}

impl Parse for TypeParam {
    fn parse(input: ParseStream) -> Result<Self> {
        let attrs = input.call(Attribute::parse_outer)?;
        let ident = input.parse()?;

        Ok(TypeParam {
            attrs,
            ident,
        })
    }
}
```

which results in:

```
error[E0284]: type annotations needed: cannot satisfy `<std::result::Result<syn::Ident, syn::Error> as std::ops::Try>::Residual == <std::result::Result<_, syn::Error> as std::ops::Try>::Residual`
  --> src/lib.rs:13:34
   |
13 |         let ident = input.parse()?;
   |                                  ^ cannot satisfy `<std::result::Result<syn::Ident, syn::Error> as std::ops::Try>::Residual == <std::result::Result<_, syn::Error> as std::ops::Try>::Residual`
```

### PR DESCRIPTION
**Pull request by lcnr**
The basic idea of this PR is that we don't generalize aliases when the instantiation could fail later on, either due to the *occurs check* or because of a universe error. We instead replace the whole alias with an inference variable and emit a nested `AliasRelate` goal. This `AliasRelate` then fully normalizes the alias before equating it with the inference variable, at which point the alias can be treated like any other rigid type.

We now treat aliases differently depending on whether they are *rigid* or not. To detect whether an alias is rigid we check whether `NormalizesTo` fails. While we already do so inside of `AliasRelate` anyways, also doing so when instantiating a query response would be both ugly/difficult and likely inefficient. To avoid that I change `instantiate_and_apply_query_response` to relate types completely structurally. This change generally removes a lot of annoying complexity, which is nice. It's implemented by adding a flag to `Equate` to change it to structurally handle aliases.

We currently always apply constraints from canonical queries right away. By providing all the necessary information to the canonical query, we can guarantee that instantiating the query response never fails, which further simplifies the implementation. This does add the invariant that *any information which could cause instantiating type variables to fail must also be available inside of the query*.

While it's acceptable for canonicalization to result in more ambiguity, we must not cause the solver to incompletely structurally relate aliases by erasing information. This means we have to be careful when merging universes during canonicalization. As we only generalize for type and const variables we have to make sure that anything nameable by such a type or const variable inside of the canonical query is also nameable outside of it. Because of this we both stop merging universes of existential variables when canonicalizing inputs, we put all uniquified regions into a higher universe which is not nameable by any type or const variable.

I will look into always replacing aliases with inference variables when generalizing in a later PR unless the alias references bound variables. This should both pretty much fix https://github.com/rust-lang/trait-system-refactor-initiative/issues/4. This may allow us to merge the universes of existential variables again by changing generalize to not consider their universe when deciding whether to generalize aliases. This requires some additional non-trivial changes to alias-relate, so I am leaving that as future work.   

Fixes https://github.com/rust-lang/trait-system-refactor-initiative/issues/79. While it would be nice to decrement universe indices when existing a `forall`, that was surprisingly difficult and not necessary to fix this issue. I am really happy with the approach in this PR think it is the correct way forward to also fix the remaining cases of https://github.com/rust-lang/trait-system-refactor-initiative/issues/8.

### CODE REVIEW

**Comment by compiler-errors** [comment]
gna mark this as author until it's out of draft?

@rustbot author

File: compiler/rustc_trait_selection/src/solve/eval_ctxt/canonical.rs [hunk 1]
```diff
@@ -293,32 +293,37 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
...
-                    })
-                    .map_err(|e| {
-                        debug!(?e, "failed to equate");
-                        NoSolution
-                    })?,
-            );
+            let trace = ToTrace::to_trace(&cause, true, orig, response);
+            infcx
+                .combine_fields(trace, param_env, DefineOpaqueTypes::No)
+                .equate(true)
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/solve/eval_ctxt/canonical.rs:322]: can this just be `.eq` still?

**Comment by lcnr** [comment]
blocking this on #120549, It needs alias relate to not use structurally relate for the outermost alias which is hard to implement as long as we still handle opaque types there as well

File: tests/ui/traits/next-solver/normalize-unsize-rhs.rs [hunk 2]
```diff
@@ -1,5 +1,6 @@
 // compile-flags: -Znext-solver
-// check-pass
```
Comment (lcnr) [inline tests/ui/traits/next-solver/normalize-unsize-rhs.rs:2]: have to figure out why this fails now :/ do we not `select_where_possible` at some important point? probably

Comment (compiler-errors) [reply tests/ui/traits/next-solver/normalize-unsize-rhs.rs:2]: Why would this turn into `non-primitive cast` error due to a missing `select_where_possible`?

**Pull request merged**